File: backend/app/schemas/book.py
from datetime import date, datetime

from pydantic import BaseModel


class TagOut(BaseModel):
    id: int
    name: str

    model_config = {"from_attributes": True}


class BookListItem(BaseModel):
    id: int
    title: str
    title_sort: str | None = None
    author: str | None = None
    publisher: str | None = None
    publish_date: date | None = None
    isbn: str | None = None
    language: str | None = None
    cover_key: str | None = None
    has_epub: bool = False
    rating: int = 0
    series_name: str | None = None
    series_index: float | None = None
    tags: list[TagOut] = []

    model_config = {"from_attributes": True}


class BookDetail(BookListItem):
    calibre_id: int | None = None
    uuid: str | None = None
    author_sort: str | None = None
    google_id: str | None = None
    amazon_id: str | None = None
    description: str | None = None
    epub_key: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class BookCreate(BaseModel):
    title: str
    title_sort: str | None = None
    author: str | None = None
    author_sort: str | None = None
    publisher: str | None = None
    publish_date: date | None = None
    isbn: str | None = None
    google_id: str | None = None
    amazon_id: str | None = None
    language: str | None = "eng"
    description: str | None = None
    rating: int = 0
    series_name: str | None = None
    series_index: float | None = None
    tags: list[str] = []


class BookUpdate(BaseModel):
    title: str | None = None
    title_sort: str | None = None
    author: str | None = None
    author_sort: str | None = None
    publisher: str | None = None
    publish_date: date | None = None
    isbn: str | None = None
    google_id: str | None = None
    amazon_id: str | None = None
    language: str | None = None
    description: str | None = None
    rating: int | None = None
    series_name: str | None = None
    series_index: float | None = None
    tags: list[str] | None = None


class PaginatedBooks(BaseModel):
    items: list[BookListItem]
    total: int
    page: int
    per_page: int
    pages: int

File: frontend/src/types/book.ts
export interface Tag {
  id: number;
  name: string;
}

export interface TagWithCount extends Tag {
  count: number;
}

export interface BookListItem {
  id: number;
  title: string;
  title_sort: string | null;
  author: string | null;
  publisher: string | null;
  publish_date: string | null;
  isbn: string | null;
  language: string | null;
  cover_key: string | null;
  has_epub: boolean;
  rating: number;
  series_name: string | null;
  series_index: number | null;
  tags: Tag[];
}

export interface BookDetail extends BookListItem {
  calibre_id: number | null;
  uuid: string | null;
  author_sort: string | null;
  google_id: string | null;
  amazon_id: string | null;
  description: string | null;
  notes: string | null;
  epub_key: string | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface PaginatedBooks {
  items: BookListItem[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
}

export interface BookSearchParams {
  q?: string;
  author?: string;
  tags?: string;
  year_from?: number;
  year_to?: number;
  rating_min?: number;
  has_epub?: boolean;
  sort?: string;
  order?: "asc" | "desc";
  page?: number;
  per_page?: number;
}

export type ViewMode = "grid" | "list" | "masonry" | "table";
